### PR TITLE
Capture information for additional query types and aggregations types

### LINKED ISSUE #11364: [Search Query Categorization] Expanding the capture of query type and level information
In Phase 1 of Search Query Categorization we captured query type and level information for a few query types.

In Phase 2 as part of this task we need to expand this to capture more query types and try to include all the aggregation query types.

After phase 2 we will be capturing query type and level information for the following query types (Including phase 1 and 2):
BoolQuery
BoostingQuery
ConstantScoreQuery
DisMaxQuery
DistanceFeatureQuery
ExistsQuery
FieldMaskingSpanQuery
FuzzyQuery
GeoBoundingBoxQuery
GeoDistanceQuery
GeoPolygonQuery
GeoShapeQuery
IntervalQuery
MatchAllQuery
MatchPhraseQuery
MatchQuery
MultiMatchQuery
PrefixQuery
QueryStringQuery
RangeQuery
RegexpQuery
ScriptQuery
SimpleQueryString
TermQuery
WildcardQuery
FunctionScoreQuery

We will be incrementing counters for each of these query types and capturing the level information as a tag to the counters. Counters will be incremented using the Telemetry metrics framework.

Phase 1: https://github.com/opensearch-project/OpenSearch/issues/10250
Phase 2 Meta issue : https://github.com/opensearch-project/OpenSearch/issues/11040
Query Visibility Roadmap : https://github.com/opensearch-project/OpenSearch/issues/11008

### PR DESCRIPTION
**Pull request by deshsidd**
### Description
This changes is part of the [Query Visibility Project](https://github.com/opensearch-project/OpenSearch/issues/11008) and [Search Query Categorization phase 2](https://github.com/opensearch-project/OpenSearch/issues/11040) initiative.

In this change we address the following:
[Search Query Categorization] Expanding the capture of query type and level information #11364
[Search Query Categorization] Identifying the types of aggregations #11366

### Related Issues
Resolves #11364,#11366

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/action/search/SearchQueryAggregationCategorizer.java
Comment (ansjcy) [inline server/src/main/java/org/opensearch/action/search/SearchQueryAggregationCategorizer.java:97]: Hmm, I'm not sure if it is a good practice to write so many verbose if else statements. There might be better ways to achieve this.

For example, we can create a static mapping class (or use a singleton) to record the builder to tag mapping, and use a for loop here to have better readability.

Or, should we consider making the tag as an attribute of the aggregation builder itself, and we can create a generic function in the aggregation builder, so that we can directly add the corresponding counter based on tag.

File: server/src/main/java/org/opensearch/action/search/SearchQueryCategorizingVisitor.java
Comment (ansjcy) [inline server/src/main/java/org/opensearch/action/search/SearchQueryCategorizingVisitor.java:82]: Same as above. We should consider improving this part of logic.

**Comment by ansjcy** [comment]
Should we consider moving the categorizer logic to the query insight plugin as part of the phase 2 effort? rather than keeping adding logic in the core search workflow.

**Comment by ansjcy** [comment]
LGTM! The only concern I have is we are adding addtional logic that might have an impact on resource usage and latency in core search path. Let's make sure to have enought load test coverage to understand the impact, and create a task to refactor this to an async workflow once the query insight framework is merged in. (https://github.com/opensearch-project/OpenSearch/pull/11506)

**Review by ansjcy — APPROVED** [review]

File: server/src/main/java/org/opensearch/action/search/SearchQueryCategorizingVisitor.java
Comment (jainankitk) [inline server/src/main/java/org/opensearch/action/search/SearchQueryCategorizingVisitor.java:102]: Nice refactoring! Although, it will be even to encapsulate this information within `SearchQueryCounters`. The visitor can just pass the query builder class and `SearchQueryCounters` can make the decision to increment the corresponding counter. That will also ensure that these counters are not public

File: server/src/main/java/org/opensearch/action/search/SearchQueryCounters.java
Comment (jainankitk) [inline server/src/main/java/org/opensearch/action/search/SearchQueryCounters.java:42]: The naming for these query counters is inconsistent. We should stick to without Query suffix nomenclature. Few of the violations are `queryStringQueryCounter`(additional Query), `boostCounter`(missing ing), `regexCounter` (missing p)

File: server/src/main/java/org/opensearch/action/search/SearchQueryCounters.java
Comment (jainankitk) [inline server/src/main/java/org/opensearch/action/search/SearchQueryCounters.java:68]: boost?

File: server/src/main/java/org/opensearch/action/search/SearchQueryCounters.java
Comment (jainankitk) [inline server/src/main/java/org/opensearch/action/search/SearchQueryCounters.java:73]: Are we using camel case here? If yes, this and few other places should be changed. If not, others like `disMax` should be changed

File: server/src/main/java/org/opensearch/index/query/FieldMaskingSpanQueryBuilder.java
Comment (jainankitk) [inline server/src/main/java/org/opensearch/index/query/FieldMaskingSpanQueryBuilder.java:59]: Any specific reason for this refactoring?

File: server/src/main/java/org/opensearch/action/search/SearchQueryAggregationCategorizer.java
Comment (jainankitk) [inline server/src/main/java/org/opensearch/action/search/SearchQueryAggregationCategorizer.java:66]: The signature of this handler map can just be `Map<Class<? extends AggregationBuilder>, String>`. Why do you need the function?

File: server/src/main/java/org/opensearch/action/search/SearchQueryAggregationCategorizer.java
Comment (jainankitk) [inline server/src/main/java/org/opensearch/action/search/SearchQueryAggregationCategorizer.java:62]: `initializeAggregationHandlers` should be void return type. It can directly initialize the instance variable, or initialize before invocation and let the method just fill the handlers map

File: server/src/main/java/org/opensearch/action/search/SearchQueryCategorizingVisitor.java
Comment (jainankitk) [inline server/src/main/java/org/opensearch/action/search/SearchQueryCategorizingVisitor.java:66]: This method should have void return type

File: server/src/main/java/org/opensearch/index/query/FieldMaskingSpanQueryBuilder.java
Comment (deshsidd) [reply server/src/main/java/org/opensearch/index/query/FieldMaskingSpanQueryBuilder.java:59]: Need the NAME attribute for my code to work. Also good to follow convention of defining the NAME for all the *QueryBuilders

**Review by jainankitk — APPROVED** [review]
Thanks for addressing the review comments! LGTM!

**Review by msfroh — APPROVED** [review]

**Comment by msfroh** [comment]
@peternied  -- This one seems to be stuck on "Minimum approval count", though I've approved it. 

Do you know if there's something else that we need to do?

**Comment by peternied** [comment]
@msfroh Can you dismiss and re-review?

**Review by msfroh — APPROVED** [review]

**Pull request merged**
